2814: Integrate new dashboard structure into tupaia.org

chris-bes opened the pull request:
### Issue: https://github.com/beyondessential/tupaia-backlog/issues/2814

### Changes:

- Integrate new dashboard structure into tupaia.org

### TODO:
- Admin panel tabs for new tables https://github.com/beyondessential/tupaia-backlog/issues/2939

chris-bes commented on packages/lesmis-server/src/routes/ReportRoute.ts:
```diff
@@ -26,20 +27,23 @@ export class ReportRoute extends Route {
       case 'dashboard': {
         if (legacy === 'true') {
-          const legacyReport = await this.webConfigConnection.fetchDashboardReport({
+          const legacyReport = await this.webConfigConnection.fetchDashboardReport(reportCode, {
```
A bit awkward here because web-config-server report endpoint will also check for `legacy`. But I'm not too fussed about it 😬

chris-bes commented on packages/web-config-server/src/apiV1/RouteHandler.js:
```diff
@@ -54,10 +61,17 @@ export class RouteHandler {
+      });
+    }
+
+    return this.project;
```
We need project in order to perform permission check for `report` endpoint

chris-bes commented on packages/web-config-server/src/apiV1/index.js:
```diff
@@ -56,14 +55,13 @@ export const getRoutesForApiV1 = () => {
   api.get('/organisationUnitSearch', handleWith(OrgUnitSearchHandler));
-  api.get('/dashboard', handleWith(DashBoardHandler)); // Legacy style dashboards
-  api.get('/view', handleWith(ViewHandler));
+  api.get('/dashboard', handleWith(DashboardHandler)); // Legacy style dashboards
```
I will remove this endpoint and also drop `dashboardReport` + `dashboardGroup` in the next PR (and also add Admin Panel tabs for the new tables)

chris-bes commented on packages/web-config-server/src/apiV1/report.js:
```diff
@@ -0,0 +1,123 @@
+    };
+  };
+
+  async buildReportData(reportCode) {
```
I actually haven't tested this route with any new reports yet. Solutions team implemented some new `reportServerDataBuilder` dashboards but they haven't been tested. Once they're tested, I can migrate them to use the new structure to test if this is working

EMcQ-BES replied:
Oh I have a test report I've been using, so I ran your branch to see if it works.

![image](https://user-images.githubusercontent.com/69437175/122328374-a6246300-cf83-11eb-810d-1a560113dfe6.png)

chris-bes commented on packages/web-frontend/src/components/View/SingleValueWrapper.js:
```diff
@@ -26,14 +26,16 @@ import { formatDataValue } from '../../utils';
   render() {
-    const { name, valueType, value, total, value_metadata } = this.props.viewContent;
-    const metadata = value_metadata || this.props.viewContent[`${name}_metadata`];
+    const { name, valueType, data, total, value_metadata: valueMetadata } = this.props.viewContent;
```
Single value reports no longer returns just `value`. I changed this I can remove `viewJson` from the `report` endpoint, and also it can be consistent with how other dashboard item types receiving data

chris-bes commented on packages/web-frontend/src/containers/DashboardGroup/index.js:
```diff
@@ -39,31 +39,24 @@ export default class DashboardGroup extends Component {
-      if (!view.requiresDataFetch) {
+    const infoViews = tab.items.map(view => {
+      const uniqueViewId = getUniqueViewId(organisationUnitCode, dashboardCode, view.code);
+      if (view.noDataFetch) {
```
This was determinded in the view.js endpoint before by checking if `dataBuilder` is null. Now I moved it to dashboard_item config

chris-bes commented on packages/web-frontend/src/screens/mobile/HomeScreen/index.js:
```diff
@@ -16,7 +15,7 @@ import { Dashboard } from '../../../components/mobile/Dashboard';
 import { WHITE } from '../../../styles';
 import {
-  selectCurrentDashboardGroupCode,
+  selectCurrentDashboardName,
```
So before, in a lot of places, the variable/method names are often `dashboardGroupCode`, even though they are `dashboardName` instead of codes. I found it a bit confusing. So I tried to change the URL to used `dashboardCode` instead of `dashboardName`, but this is tricky because in a lot of places we only have access to `dashboardName` (eg: project default project is reference by `dashboardName`). So I went with changing the variable/method names from `dashboardGroupCode` to `dashboardName` instead to avoid confusion

chris-bes commented:
One thing I want to ask is, with the new dashboards endpoint. If a dashboard is empty, it won't be returned to the client. However, in LESMIS, there are a few dashboards that are empty. (screenshot below)
![image](https://user-images.githubusercontent.com/63621318/122132842-f239af80-ce7e-11eb-9d04-1671a1ea0faa.png)

I don't know if we have to maintain the empty dashboards? If for any reason we have to, then a solution is to add an empty dashboard_item in these until we have real reports?

chris-bes commented on packages/web-config-server/src/apiV1/dashboards.js:
```diff
@@ -3,33 +3,64 @@
+
+  getDashboardsWithItems = async (dashboards, entity, projectCode) => {
+    const permissionGroups = await this.req.getUserGroups(entity.code);
     const sortedDashboards = orderBy(dashboards, ['sort_order', 'name']);
```
Is it okay to sort all dashboards by name? The order will be different from how it is now in tupaia.org

chris-bes commented:
One thing I want to ask is, with the new dashboards endpoint. If a dashboard is empty, it won't be returned to the client. However, in LESMIS, there are a few dashboards that are empty. (screenshot below)
![image](https://user-images.githubusercontent.com/63621318/122132842-f239af80-ce7e-11eb-9d04-1671a1ea0faa.png)

I don't know if we have to maintain the empty dashboards? If for any reason we have to, then a solution is to add an empty dashboard_item in these until we have real reports?

EMcQ-BES commented on packages/web-frontend/src/containers/EnlargedDialog/EnlargedDialog.js:
```diff
@@ -93,6 +96,13 @@ const EnlargedDialogComponent = ({
   // Regardless of the drillDown level, we pass the base view content through
   const baseViewContent = contentByLevel?.[0]?.viewContent;
   const drillDownContent = contentByLevel?.[drillDownState.drillDownLevel]?.viewContent;
+  const baseViewConfig = contentByLevel?.[0]?.viewConfig;
+  const drillDownConfig = contentByLevel?.[drillDownState.drillDownLevel]?.viewConfig;
```
I'm not entirely sure how `contentByLevel` is being calculated? Does that need to be updated for the new way drill downs are laid out?

chris-bes replied:
Yep so it stores the viewConfig/viewContent of each level of drill down.

EMcQ-BES reviewed (commented):
Looks great Chris. A couple of questions but those are mostly for my benefit since I'm not super familiar with the web-frontend code

Looks like you there's only a couple of places with big changes and the rest is just updating variable names and passing data through the whole chain.

edmofro commented:
> One thing I want to confirm is, with the new dashboards endpoint, if a dashboard is empty, it won't be returned to the client. However, in LESMIS, there are a few dashboards that are empty. (screenshot below)
> ![image](https://user-images.githubusercontent.com/63621318/122132842-f239af80-ce7e-11eb-9d04-1671a1ea0faa.png)
> 
> I don't know if we have to maintain the empty dashboards? If for any reason we have to, then a solution is to add an empty dashboard_item in these until we have real reports?

Thanks for pointing this out @chris-bes. There was a reason - it's how UNICEF wanted it. Not sure if it's still important, and have asked Erin [here](https://github.com/beyondessential/tupaia/pull/2673#issuecomment-861884177)

chris-bes commented on packages/web-frontend/src/containers/DashboardItem/index.js:
```diff
@@ -54,41 +54,37 @@ export class DashboardItem extends Component {
     if (
       nextProps.infoViewKey !== this.props.infoViewKey ||
-      this.props.organisationUnit !== nextProps.organisationUnit ||
-      this.props.disasters !== nextProps.disasters
```
It is actually redundant because this component never receive any `disasters` props from else where as I've checked.

chris-bes commented:
@EMcQ-BES Sorry for addressing this PR quite late. I've pushed 2 commits:
https://github.com/beyondessential/tupaia/pull/2673/commits/c74e5188cc8fbd6f44df3c64ddde7d270d6cc5d9: Update new drill down + some fixes for some bugs that I've just found
https://github.com/beyondessential/tupaia/pull/2673/commits/bca092d03898ad78526980f6d15ab9583bec7b29: Addressed your review comments from both this PR and the former PR.

chris-bes commented on packages/database/src/migrations/20210625012001-FixCustomLegacyReports-modifies-data.js:
```diff
@@ -0,0 +1,230 @@
+      leftColumnHeader: 'Normal',
+      rightColumnHeader: 'Current',
+      dataPairNames: ['Inpatient beds'],
+    },
```
A bit of context, there are actually a few legacy data builders that reference `viewJson`. They have access to it because the previous `view` endpoint queries it. With the new `report` endpoint which should only reference either the modern report or the legacy report tables, it no longer queries for `viewJson` (now called `config` in dashboard_item). 

Most of the `viewJson` references in these legacy data builders are just hardcoded strings which is required to calculate the data. So what I did here is just move whatever they were referencing from `viewJson` before to `data_builder_config`. It is not ideal, but I feel that since these data builders are legacy, I really don't think it's worth to spend time to refactor them to fit the new endpoint structure.

chris-bes commented on packages/web-config-server/src/apiV1/dataBuilders/generic/comparison/compareDataElementPairs.js:
```diff
@@ -18,13 +18,17 @@ import flatten from 'lodash.flatten';
  */
 export const compareDataElementPairs = async (
-  { dataBuilderConfig, viewJson, query },
+  { dataBuilderConfig, query },
```
removing `viewJson` reference from legacy data builders like I explained above

edmofro commented:
@EMcQ-BES according to https://github.com/beyondessential/tupaia/pull/2673#issuecomment-868173347 just https://github.com/beyondessential/tupaia/commit/bca092d03898ad78526980f6d15ab9583bec7b29 left to review - if you can do that we can get this whole load of stuff into regression testing!

EMcQ-BES commented:
[bca092d](https://github.com/beyondessential/tupaia/commit/bca092d03898ad78526980f6d15ab9583bec7b29) looks perfectly good to me, the fixes it makes are mostly for readability rather than functionality

EMcQ-BES reviewed (approved):
I admittedly haven't re-looked at everything, but the changes since my previous review all look good to me

chris-bes commented:
I have to admit that I'm not very confident with this fix since I'm not familiar with how styling in mobile works. But I have done some testing and it seems fine so far https://github.com/beyondessential/tupaia/pull/2673/commits/f51dc58045b4a7f27d30b6893377a10c1e9e0468

edmofro commented:
@chris-bes your fix looks good (both @tcaiger and I had a look). That sneaky mv fast refresh change is back, but I'll get that sorted now.

chris-bes commented:
@tcaiger https://github.com/beyondessential/tupaia/pull/2673/commits/ae3dfa12e4924a6ae46cc0f5818f756274e6c28c LGTM!
